src/docfx/build/xref/InternalXrefMap.cs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class InternalXrefMap
+    {
+        private readonly IReadOnlyDictionary<string, XrefSpec> _map;
+
+        private InternalXrefMap(IReadOnlyDictionary<string, XrefSpec> map)
+        {
+            _map = map;
+        }
+
+        public static async Task<InternalXrefMap> Create(Context context, IEnumerable<Document> files)
+        {
+            ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefConflicts = new ConcurrentDictionary<string, ConcurrentBag<XrefSpec>>();
+            ConcurrentDictionary<string, XrefSpec> xrefs = new ConcurrentDictionary<string, XrefSpec>();
+            Debug.Assert(files != null);
+            using (Progress.Start("Building Xref map"))
+            {
+                await ParallelUtility.ForEach(files.Where(f => f.ContentType == ContentType.Page), file => Load(context, xrefs, xrefConflicts, file), Progress.Update);
+                HandleXrefConflicts(context, xrefs, xrefConflicts);
+            }
+            return new InternalXrefMap(xrefs);
+        }
+
+        public bool Resolve(string uid, out XrefSpec xref) => _map.TryGetValue(uid, out xref);
+
+        private static Task Load(Context context, ConcurrentDictionary<string, XrefSpec> xrefs, ConcurrentDictionary<string, ConcurrentBag<XrefSpec>> xrefConflicts, Document file)
+        {
+            try
+            {
+                var errors = new List<Error>();
+                var content = file.ReadText();
+                if (file.FilePath.EndsWith(".md", PathUtility.PathComparison))
+                {
+                    TryAddXref(xrefs, xrefConflicts, LoadMarkdown(content, file));
+                }
+                else if (file.FilePath.EndsWith(".yml", PathUtility.PathComparison))
+                {
+                    var (yamlErrors, token) = YamlUtility.Deserialize(content);
+                    errors.AddRange(yamlErrors);
+                    TryAddXref(xrefs, xrefConflicts, LoadSchemaDocument(errors, token, file));
+                }
+                else if (file.FilePath.EndsWith(".json", PathUtility.PathComparison))
+                {
+                    var (jsonErrors, token) = JsonUtility.Deserialize(content);
+                    errors.AddRange(jsonErrors);
+                    TryAddXref(xrefs, xrefConflicts, LoadSchemaDocument(errors, token, file));
+                }
```

Comment:
need to exclude toc(.experimental).{yml|json}

Reply:
only `ContentType.Page` would be loaded here, toc would be `ContentType.TableOfContents`, right?

---
In reply to: [219100080](https://github.com/dotnet/docfx/pull/3437#discussion_r219100080) [](ancestors = 219100080)

Reply:
Yes, you are right.

---
In reply to: [219362884](https://github.com/dotnet/docfx/pull/3437#discussion_r219362884) [](ancestors = 219362884,219100080)